src/kernel/merkletree.h

```diff
@@ -19,25 +34,39 @@ namespace CryptoKernel {
             MerkleNode(const BigNum& left);
             
             static std::shared_ptr<MerkleNode> makeMerkleTree(const std::set<BigNum>& leaves);
-        
+            static std::shared_ptr<CryptoKernel::MerkleNode> makeMerkleTreeFromProof(std::shared_ptr<CryptoKernel::MerkleProof> proof);
+            std::shared_ptr<CryptoKernel::MerkleProof> makeProof(BigNum proofValue);
             BigNum getMerkleRoot() const;
             
             BigNum getLeftVal() const;
             BigNum getRightVal() const;
-            
+            std::shared_ptr<MerkleNode>  getLeftNode();
+            std::shared_ptr<MerkleNode>  getRightNode();
+            MerkleNode* getAncestor();
+
         private:
-            bool leaf;
+            MerkleNode* ancestor;
             
             std::shared_ptr<MerkleNode> leftNode;
             std::shared_ptr<MerkleNode> rightNode;
-            
+
             BigNum leftVal;
             BigNum rightVal;
-            
-            BigNum root;
-            
+                        
+            CryptoKernel::MerkleNode* findDescendant(BigNum needle);
```

Comment:
`needle` should be both `const` and passed by reference. I think the function itself can be marked `const` too.